clang/lib/CIR/Lowering/ThroughMLIR/LowerCIRLoopToSCF.cpp

```diff
@@ -261,6 +274,40 @@ void SCFWhileLoop::transferToSCFWhileOp() {
   rewriter->eraseBlock(&scfWhileOp.getAfter().back());
 }
 
+void SCFDoLoop::transferToSCFWhileOp() {
+  // only support a simple do-while
+  // FIXME: can not support nested do-while
+
+  auto scfWhileOp = rewriter->create<mlir::scf::WhileOp>(
+      DoOp.getLoc(), DoOp->getResultTypes(), adaptor.getOperands());
+
+  rewriter->createBlock(&scfWhileOp.getBefore());
+  rewriter->createBlock(&scfWhileOp.getAfter());
+
+  rewriter->cloneRegionBefore(DoOp.getBody(), &scfWhileOp.getBefore().back());
+  rewriter->eraseBlock(&scfWhileOp.getBefore().back());
+
+  rewriter->cloneRegionBefore(DoOp.getCond(), &scfWhileOp.getAfter().back());
+  rewriter->eraseBlock(&scfWhileOp.getAfter().back());
+
+  rewriter->inlineBlockBefore(&scfWhileOp->getRegion(1).back(),
+                              &scfWhileOp->getRegion(0).back(),
+                              scfWhileOp->getRegion(0).back().end());
```

Comment:
It looks like you placed the condition block at the end of the After region, and then moved the condition block to the Before region. Why not complete this in one step by placing the condition block directly at the end of the Before region?

Reply:
Sorry, I tried the way you siad but got erro info before. Maybe I need to learn MLIR API deeply.

Reply:
OK, I'm not very familiar with the SCF dialect, but it seems that a simple do-while loop should not pose any problems (though nesting other control flow statements might cause issues). Maybe we can see what other reviewers think.